### ISSUE TITLE
RFC: Renderer redesign

### ISSUE THREAD

**Issue by nchashch** [amethyst/amethyst#159]
# Current state of renderer:
At the end of the frame all `Renderable`s and `Light`s are iterated over and packed into a `Scene`, which is then submitted to `amethyst_renderer::Renderer` along with `Pipeline` struct (it contains all the information about passes/targets), which produces an `Encoder` and then `encoder.flush(&mut device)` is called. So the work of producing an `Encoder` isn't parallelized.

# Proposal:

## Compile time backend choice (or @msiglreith's proposed way of having runtime choice without wrappers)
Dealing with rendering this way would be easier if there are no `GfxDeviceInner`, `EncoderInner`, `MainTargetInner`, etc. So code examples in this issue use `gfx_types` module defined in #155.

### _Update_: Problems with trait object `Component`s which make using them unacceptable
- It breaks data locality (not idiomatic ECS design/bad for cache friendliness)
- It introduces additional levels of indirection (a `Box`ed value/virtual methods) which have a performance cost (it is significant because `Renderable`s are iterated over/dereferenced every frame and there can be a large number of `Renderable`s in the `World`)

Define a `Renderable` as something which can fill an `Encoder`, so the most natural thing would be to make `Renderable` a trait implementing a method for filling an `Encoder`.

Maybe it would make sense to store a separate `Pipeline` for each `Renderable` (to make `Renderable`s self contained and allow for different graphic pipelines for different `Renderables`).
```rust
pub trait Renderable {
    fn render_to_encoder(&self,
                         world: &specs::World,
                         main_target: &MainTarget,
                         encoder: &mut gfx::Encoder<gfx_types::Resources, gfx_types::CommandBuffer>);
}
```
Then a component trait can be implemented for a boxed `Renderable` trait object.
```rust
impl Component for Box<Renderable + Send + Sync> {
    type Storage = VecStorage<Box<Renderable + Send + Sync>>;
}
```
Different concrete structs can now implement this trait, for example:
```rust
pub struct StaticModel {
    pub mesh: Mesh,
    pub ka: Texture,
    pub kd: Texture,
    pub pipeline: Pipeline,
}

impl Renderable for StaticModel { ... }
```
And they can be attached to entities like this:
```rust
let static_model = asset_manager.create_static_model("mesh_name", "ka_name", "kd_name");
let static_model = Box::new(static_model);
world.create_now().
         .with::<Box<Renderable + Send + Sync>>(static_model);
```

## Rendering processor
Main thread owns `Factory` and `Device` gfx structs. `Encoder`s are constructed on the main thread and sent using a channel to `RenderingProcessor` (like in [peagasus](https://github.com/kvark/pegasus)).

Processor recieves `Encoder`s from the main thread and then iterates over all `Box<Renderable + Send + Sync>`s (maybe using something like [rayon](https://github.com/nikomatsakis/rayon) for parallelism), calls `render_to_encoder` on each one, and sends filled `Encoder`s back to the main thread (using another channel), then `Encoder`s are recieved on the main thread iterated over and `encoder.flush(&mut device)` is called for each one.

`Clear` pass is performed before all the encoders are flushed, and can't be invoked inside `render_to_encoder` implementation (otherwise only one `Renderable` will apear on the screen).

I have so far written a version of this without any parallelism or modifications to `amethyst_renderer` (a separate `Encoder` is filled and flushed for each `Renderable` on the main thread and there is no rendering processor).

## Benefits:
- More flexibility: many different kinds of `Renderable`s can be implemented without touching the rendering processor: `Spline`s, `StaticModel`s, `DynamicModel`s, `Sprite`s, something custom.
- Parallelism: `Encoder`s are filled in parallel (so there is a possible performance improvement).

## Costs:
- Complexity/ergonomics cost.
- `amethyst_renderer` needs to be significantly modified.
- There is no performance benefit for single threaded platforms (only additional overhead).

## Unsolved problems:
- Will this approach yield a performance improvement over just doing it in a single thread (there is additional overhead of creating and sending a ton of `Encoder`s)?
- How to construct and send the right number of `Encoder`s every frame (one for each `Renderable`)?
- How to access concrete structs (outside of rendering processor) which implement `Renderable` trait hidden behind a trait object? (for example to change a texture for a `StaticModel` or to switch an animation for a `DynamicModel`).

**Comment by palodequeso** [comment]
VR is mentioned in https://github.com/amethyst/amethyst/issues/160 as simply working with inputs, but OpenVR has API methods you have to call to submit both eye frame buffers at the end, so that would have to tie into the renderer as well. I know it's not on the short term road-map, but perhaps it's worth having a tid-bit in here talking about how it could be handled?

**Comment by nchashch** [comment]
I guess it would involve special render targets for left/right eye. The renderer would have to render the scene twice per frame (for each eye) at 90 FPS with all the VR specific things (like software optics and color correction), which I think can be done without significantly changing the design of the renderer. Then these lef/right eye targets can be submitted using OpenVR API at the end of the frame in the main thread.

**Comment by palodequeso** [comment]
Yea, but if I'm not mistaken, the API can do all of the color correction and optical bending and what not, you just have to submit the two eye targets? I'll have to double check. Does the renderer currently have the ability to define the final render targets? If so, then it should be good to go as you've said.

**Comment by nchashch** [comment]
I could render the scene to custom render targets using current `amethyst_renderer` pretty easily in #115, although some visual artifacts creeped up.

If it just requires submitting eye targets and the API can do the corrections then I guess we only will have to worry about stereoscopic rendering and performance, also it might be nice to implement some VR specific optimizations like rendering the distorted areas in lower resolution.

**Comment by palodequeso** [comment]
That seems reasonable, although, as eye tracking comes into play, that sort of thing will need to be flexible based on the hardware.

**Comment by kvark** [comment]
```rust
impl Component for Box<Renderable + Send + Sync>
```
This is not idiomatic to `specs` or ECS in general, since it removes the data locality property of these components. And given that the renderables will be touched/iterated on quite a few times per frame, I'd like to see this indirections (introduced by both boxed data and the virtual methods) removed.

The idomatic way would be - defining multiple components, each for a different renderable type.

**Comment by OvermindDL1** [comment]
> The idiomatic way would be - defining multiple components, each for a different renderable type.

Absolutely this.  In my engines I've had things like `ModelComponent` that held a model handle (pointed to the asset manager via the handle), `ParticleComponent`, etc...  In the renderer system for, say, `ModelComponent` it sorted them in to passes based on if transparent or not, etc... and did the work there.

**Comment by nchashch** [comment]
> impl Component for Box<Renderable + Send + Sync>
> 
> This is not idiomatic to specs or ECS in general, since it removes the data locality property of these components. And given that the renderables will be touched/iterated on quite a few times per frame, I'd like to see these indirections (introduced by both boxed data and the virtual methods) removed.
> 
> The idiomatic way would be - defining multiple components, each for a different renderable type.

I see, so using boxed traits as `Component`s is unacceptable. The rationale behind using trait objects was to make adding custom `Renderable` types easier (so you don't have to touch engine internals or `Encoder` sending/recieving machinery in order to add a new renderable `Component`).

# Other approaches

## Simple approaches sacrificing flexibility/extensibility:

1. Have many different renderable `Component`s and process them using a single default `System` defined in the internals. This way in order to add a custom renderable `Component` you would have to add code to this single default rendering `System`, thereby touching the internals of the engine. This way sending/revieving `Encoder`s is done once in a single place and you don't have to worry about it. Also different types of renderable `Component`s won't be processed in parallel
by default (if it isn't specifically done inside of this single `System`).

2. Have many different renderable `Component`s and a `System` for each one type, each dealing with `Encoder` sending/recieving. So in order to add a renderable `Component` you would have to implement a `System` for it, which would have to send/recieve `Encoder`s (code duplication/ergonomics cost). This approach avoids touching engine internals, and there is parallel processing of different renderable `Component` types by default.

## A more idiomatic (than trait objects) approach, preserving data locality and flexibility/extensibility:

- Add a `Renderable` `Component` owning an `Option<Encoder>`.
- Implement a `System` which would supply `Renderable`s which hold `None`s with `Some<Encoder>`s (it would probably introduce a 1 frame delay between `Renderable` creation and the thing being rendered).
- At the end of the frame in the main thread iterate over all `Renderable`s owning `Some<Encoder>`s and flush them.
- Implement separate `System`s for custom renderable `Component`s (`Spline`, `StaticModel`, `DynamicModel`, etc), which would iterate over for example `(&StaticModel, &mut Renderable)` pairs and fill `Some<Encoder>`s owned by `Renderable`s based on data in `StaticModel`s.

This way if you are implementing a custom renderable `Component` you wouldn't have to worry about sending `Encoder`s between main thread/system thread and you wouldn't have to touch engine internals. You still would have to worry about data parallelism (if for example you wanted to process `(&StaticModel, &mut Renderable)` pairs in parallel).


# Thoughts on providing the right number of `Encoder`s (second problem on the unsolved problems list):

Have an `EncoderProvider` world resource, which would give out a requested number of `Encoder`s constructed on the main thread in bulk. Also this way we don't have to use channels.

**Comment by kvark** [comment]
Thanks for the update @nchashch !
> Add a Renderable Component owning an Option<Encoder>

That scares me. `Encoder` is a relatively fat object, and it should not be bundled with each component instance.

I like both variants in the "Simple approaches sacrificing flexibility/extensibility" section though.

**Comment by OvermindDL1** [comment]
For note, instead of trying to have a single component hold many types of things by boxing, having lots and lots of components tends to be far better, and if you need them managed by a singular system (like for building up the render path) then adding more tag components is good.  Having a system that operates over just `ModelComponent`'s and adds another component of, say, `TransparentRendering` let's you get a separate pipeline, just have two renderer systems that are run in their different pipelines, one for with `TransparentRendering` components and another for without `TransparentRendering` components.  Just keep breaking it up.  I routinely had thousands of component types in my engine, many of them empty (just existed as the bitarray to know if it exists or not at all, just a fast flag/tag).

**Comment by kvark** [comment]
FYI, for the [Lego brick](https://users.rust-lang.org/t/game-engine-design-lego-bricks/9151/1) design that I'm experimenting with, I'm thinking of a rendering system that:
  - has a specific marker component to flag those rendering components (potentially used by multiple systems) that it should be processing
  - typed by a PSO, thus associated with a single PSO definition
  - has multiple PSO objects of this definition (e.g. different blend modes)
  - has a single camera
  - has a single framebuffer (as in - a compatible collection of targets)
  - has a separate channel with the device thread that it re-cycles the Encoders with

This way, the additional type dependency of PSO  is getting erased since the systems are boxed anyway, and the scope of each system is very well defined, allowing them to scan the entities independently and fill up the corresponding command buffers in parallel.

**Comment by OvermindDL1** [comment]
Looks good except the single camera, is there going to be no way to render difference cameras to different framebuffers?  That is pretty required for VR as one example.  Or are you going to be expected to render, move the camera (and rebuild its pipeline) and render again, then render those framebuffers to the proper areas on the screen (perhaps in yet another scene via that same camera)?

**Comment by kvark** [comment]
@OvermindDL1 first, the camera/framebuffer are the implementation detail (as I edited/appended to the post), so you should be able to extend this with no issues to the outside world.

Secondly, if you are rendering 2 eyes separately using different cameras, you might as well have 2 different intances of this system running in parallel.

**Comment by msiglreith** [comment]
Compile time backend are merged into the master branch.

I would also prefer the marker trait approach and rendering processors would then just pick the components with the needed traits.
Can we have a pool of encoders for each thread, which will be submitted to the device? This seems to be the prefered way for next gen APIs.

**Comment by nchashch** [comment]
Proposal for handling `Encoder` logistics:
1. `ecs::resource::EncoderPool`:
It would look something like this:
```rust
struct EncoderPool {
        // `Mutex` is necessary because `Encoder` doesn't implement `Sync`,
        // which is required to make something an ecs resource.
        encoders: Vec<Mutex<gfx_types::Encoder>>,
}

impl EncoderPool {
    pub fn new(num_encoders: usize, &mut gfx_types::Factory) -> EncoderPool {
        // Create `num_encoders` `Encoder`s, put them inside `EncoderPool`, and return it.
        ...
    }

    pub fn lock_encoder(&mut self) -> Option<MutexGuard<gfx_types::Encoder>> {
        // If there are still unlocked `Encoder`s, lock one and return `Some(locked_encoder)`,
        // otherwise return `None`.
        ...
    }

    pub fn flush(&mut self, &mut gfx_types::Device) {
        // Lock and flush all `Encoder`s in the pool.
        ...
    }
}
```
At initialization an `EncoderPool` with `num_encoders` equal to the number of threads in `specs` thread pool is created on main thread and added to the `specs::World` as a resource.

Every rendering thread would call `EncoderPool::lock_encoder` once to get an `Encoder` and fill it up, then when the `MutexGuard` would go out of scope the `Encoder` would be unlocked and become usable by other threads. A thread would have to wait for an `Encoder` only if all `Encoder`s in the pool are simultaneously locked (which shouldn't be a problem if number of threads = number of encoders, and every thread calls `EncoderPool::lock_encoder` only once).

At the end of the frame after all `Encoder`s were filled up, `EncoderPool::flush` is called on the main thread.

This way there is a reasonable number of `Encoder`s (equal to number of threads in `specs` thread pool) and parallel rendering can be done. Also `System`s for different renderable types can reuse this ecs resource.

**Comment by nchashch** [comment]
Thoughts on making rendering (filling up `Encoder`s) parallel in `amethyst`:

I think currently rendering can be made [task parallel](https://en.wikipedia.org/wiki/Task_parallelism) quite easily. Different rendering `System`s can lock their respective `Storage`s and run in parallel. Although it wouldn't give us any performance benefit untill we have more than one renderable `Component` type.

So in this case for example if we implemented `RenderableA`, `RenderableB`, `RenderableC` `Component`s and respective `SystemA`, `SystemB`, `SystemC` `System`s. And if we had say 750 entities, with renderable `Component`s of different types attached to entities in equal proportion (`RenderableA` for first 250, `RenderableB` for second 250, ...). We then could expect a 3 times performance increase (for rendering code) and better load distribution between cores than with a single threaded case.

The problem is that this performance benefit depends on there being many (at least as many as there are cores) renderable `Component` types and these types being attached to entities in equal proportion. Which might happen but it isn't very common. A typical game scene wouldn't have 250 `Spline`s, 250 `StaticModel`s, and 250 `DynamicModel`s. It would more likely have something like 10 `StaticModel`s for skybox/environment/gui, 7 `Spline`s for say trajectories, and 733 `DynamicModel`s for characters/grass/trees/vehicles.

A better approach would be to make rendering [data parallel](https://en.wikipedia.org/wiki/Data_parallelism). Say for example we have 1000 `RenderableA` on a 4 core machine. Then the rendering process would go like this:

1. Lock `RenderableA`s.
2. Divide `RenderableA` `Storage` into 4 equal chunks.
3. Get a `MutexGuard<Encoder>` for each of the 4 threads (using `EncoderPool`).
4. Process each `RenderableA` chunk in it's own thread (filling up 4 `Encoder`s).
5. Wait for the job to be done and `Encoder`s in the `EncoderPool` unlocked.
6. Call `EncoderPool::flush` on the main thread.

This way we would get a performance boost (over having one thread) for most cases, including cases for which task parallelism cuts it (only the order in which processing is done would change).

**Comment by msiglreith** [comment]
Thanks for writing this up! Pretty difficult to come up with an good architecture imo.

Using a global `EncoderPool` will most likely cause issues with Vulkan due to synchronization requirements for command pools and command buffers (vkspecs):
> Command pools are application-synchronized, meaning that a command pool must not be used concurrently in multiple threads. That includes use via recording commands on any command buffers allocated from the pool, as well as operations that allocate, free, and reset command buffers or the pool itself.

Therefore, I would propose one/multiple pools per thread which it's also suggested in several talks concerning this topic (e.g Vulkan Multi-Threading by Schott M.) Multiple pools per thread also allows us to wait a few frames before resetting the pool to give the queue and GPU enough time to process the cmd buffers.

This leads to specification of a certain order in which the buffers will be submitted to the queue.

Regarding the systems:
We can probably combine both, task-parallelism will drive the overall rendering pipeline (G-Buffer, light pass, shadow, AO, post-fx, etc.) to ensure the dependencies between the different stages (or maybe one system per renderpass?). Data-parallelism allows us then to produce the different secondary command buffers on multiple threads for larger renderpasses.

Further thoughts:
If we add an extra render processor in front of the actual renderer, which will preprocess the renderable components and store them separately to forward them to the actual render, we could actually start already with processing the next frame on CPU and only wait for the frame before. So we could interleave CPU and GPU tasks of 2 different frames, allowing us to better utilize the threads if for example some CPU systems are blocked due to dependencies.

**Comment by nchashch** [comment]
> Using a global EncoderPool will most likely cause issues with Vulkan due to synchronization requirements for command pools and command buffers (vkspecs):

In current Vulkan backend implementation for `gfx`, `gfx::backend::vulkan::CommandBuffer` owns both a `vk::CommandPool` and a `vk::CommandBuffer` (which was created with this command pool). So if you give a `gfx::Encoder` (which is a wrapper around a gfx `CommandBuffer`) to every thread you wouldn't reuse the same `vk::CommandPool`, since each `gfx::Encoder` would have it's own separate `vk::CommandPool`.

Here is [`gfx::backend::vulkan::command::Buffer` struct defenition](https://github.com/gfx-rs/gfx/blob/master/src/backend/vulkan/src/command.rs#L28).
And here [this `Buffer` struct is reexported as `CommandBuffer`](https://github.com/gfx-rs/gfx/blob/master/src/backend/vulkan/src/lib.rs#L27).

So it seems like if we are using `gfx` we don't have to worry about `vk::CommandPool`s, because `gfx` hides them inside `gfx::backend::vulkan::CommandBuffer`s. And giving out one `gfx::CommandBuffer` to every thread is the same as giving one `vk::CommandPool` and one `vk::CommandBuffer` to every thread (though hidden behind an abstraction).

**Comment by Xaeroxe** [comment]
@omni-viral recommended closing this, so we're closing it.

**Comment by omni-viral** [comment]
Renderer went huge redesign. It's now runs as parallel as possible. `Encoder`s are reused.
I'm closing it
